### PR TITLE
#4252 [Manager] Create workspace setting layout

### PR DESCRIPTION
**Pull request by konekoya**
### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #4531 )
- [ ] Add unit tests (or please explain why this PR is not worth having new tests)
- [x] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-manager/client/src/components/Workspace/Edit/NewEditWorkspace.js [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+];
+
+const sectionRefs = {};
+
+// Add a React ref in each item
+const listWithRefs = list.map(section => {
+  return {
+    ...section,
+    items: section.items.map(item => {
+      sectionRefs[item.text] = React.createRef();
```
Comment (konekoya) [inline ohara-manager/client/src/components/Workspace/Edit/NewEditWorkspace.js:191]: Nice trick that I learned from @eechih 💯

File: ohara-manager/client/src/context/dialog/dialogReducer.js [hunk 2]
```diff
@@ -49,7 +49,13 @@ import {
...
  */
 const initialState = {
-  dialogs: [],
+  dialogs: [
+    {
+      name: 'EDIT_WORKSPACE_DIALOG',
+      data: { tab: 'overview' },
+      createdAt: 1577681120,
+    },
+  ],
```
Comment (konekoya) [inline ohara-manager/client/src/context/dialog/dialogReducer.js:58]: @eechih I tried to enable this so it's easier to work on the edit workspace dialog. But this seems not working at all. Am I missing something here. Could you shed some light on this... 🙇

Comment (eechih) [reply ohara-manager/client/src/context/dialog/dialogReducer.js:58]: wow... api doc 沒更新到，請試試看

```
dialogs: [
  {
    name: 'editWorkspace',
    data: { tab: 'overview' },
  },
],
```

Comment (konekoya) [reply ohara-manager/client/src/context/dialog/dialogReducer.js:58]: Thanks ❤️

File: ohara-manager/client/src/components/Workspace/Settings/SettingsStyles.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    /* Offset used in scroll. We need to do this as the the dialog title height is added in the scrollIntoView API */
+    .anchor-element {
+      position: absolute;
+      margin-top: -90px;
+    }
+
+    .settings-list {
+      width: 600px;
```
Comment (eechih) [inline ohara-manager/client/src/components/Workspace/Settings/SettingsStyles.js:42]: should be 680px

**Comment by eechih** [comment]
螢幕快照 2020-04-14 上午9 12 22

螢幕寬度縮到最小時，會破圖。

File: ohara-manager/client/src/components/Workspace/Settings/SettingsStyles.js (hunk 3 above)
Comment (konekoya) [reply ohara-manager/client/src/components/Workspace/Settings/SettingsStyles.js]: I tried with 680px but it's too big and and layout breaks... Hmm, i can give it another shot :)

**Comment by konekoya** [comment]
Fixed. But was wondering why this page uses a different width and layout 🤔 I mean, could we just use a 1400px constrain and just place the menu inside it so the layout is same just like rest of fullscreen dialog?

**Comment by konekoya** [comment]
And plus, our settings are not as complex as Google's so the list seems a bit too big IMO :)

**Comment by eechih** [comment]
> And plus, our settings are not as complex as Google's so the list seems a bit too big IMO :)

主要是為了內層顯示 Table 可以有較大的空間。

File: ohara-manager/client/src/components/Workspace/Settings/data.js [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+];
+
+const sections = [
+  {
+    heading: 'Topics',
+    components: [
+      {
+        icon: <SubjectIcon />,
+        title: 'Topics in this workspace',
+        subTitle: 'I dont know what to do with this...',
```
Comment (konekoya) [inline ohara-manager/client/src/components/Workspace/Settings/data.js:97]: Testing the subtitle, will be removed when the PR is ready to merge

File: ohara-manager/client/src/components/Workspace/Settings/data.js [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+        - icon
+        - title
+        - subTitle
+        - type
+        - componentProps: 
+          - handleClick
+          - children: a valid react children
+*/
+
+const sections = [
```
Comment (konekoya) [inline ohara-manager/client/src/components/Workspace/Settings/data.js:120]: @eechih I've added a few dump component to mimic the look and feel of Chrome's settings page. I will remove these once the PR is ready to merge :)

**Comment by konekoya** [comment]
retry manager-e2e

File: ohara-manager/client/src/components/Workspace/Settings/data.js (hunk 5 above)
Comment (eechih) [reply ohara-manager/client/src/components/Workspace/Settings/data.js]: Too many commits ... difficult to review~ Can you help to squash to a commit?

**Comment by eechih** [comment]
Too many commits ... difficult to review~ Can you help to squash to a commit?

**Comment by eechih** [comment]
👍看起來很不錯，比原始設計的功能還完整。感謝 @konekoya 

有兩個小地方，可以請你再調整一下？
1. 進到 Tab Page 後，返回 Settings Page 時，Setting List 應該要保留在上次的位置。
2. 進到 Tab Page 後，點選 Settings Nav 任何一個 Item 時，應該返回到 Settings Page。

**Comment by konekoya** [comment]
> Too many commits ... difficult to review~ Can you help to squash to a commit?

Hmm, i thought it's easier to review this way. Because you can review by commit not all files. And use the files tab to do the review the other way around (review all commits at once) :)

**Comment by eechih** [comment]
> Hmm, i thought it's easier to review this way. Because you can review by commit not all files. And use the files tab to do the review the other way around (review all commits at once)

Got it, thank you.

**Comment by konekoya** [comment]
> 進到 Tab Page 後，返回 Settings Page 後，Settings List 應該保留在原本位置，不要捲到最上方。
> 進到 Tab Page 後，點選 Settings Nav 任何一個 Item 時，應該返回到 Settings Page。

Nice suggestions @eechih 😄 PR updated, please give it a look when you have the time

File: ohara-manager/client/src/components/Workspace/Settings/DeleteConfirmDialogContent.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import { Input } from '@material-ui/core';
+
+const DeleteConfirmDialogContent = ({ workspaceName, setWorkspaceName }) => {
+  return (
```
Comment (konekoya) [inline ohara-manager/client/src/components/Workspace/Settings/DeleteConfirmDialogContent.js:7]: @eechih just a component demonstrate how to disable dialog's confirm button

File: ohara-manager/client/src/components/Workspace/Settings/SettingsHooks.js [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+          - type
+          - icon: optional
+          - subTitle: optional
+          - componentProps: 
+            - handleClick: optional
+            - children: a valid react children
+*/
+
+export const useConfig = () => {
+  const [workspaceName, setWorkspaceName] = React.useState('');
```
Comment (konekoya) [inline ohara-manager/client/src/components/Workspace/Settings/SettingsHooks.js:64]: This is the easiest way that i can think of right now. The shared state can be use by both `<Dialog />` and `<DeleteConfirmDialogContent />`.

**Comment by konekoya** [comment]
@eechih I've come up with a solution to allow dialog component to enable/disable the confirm button. Please review it again. Thank you 😃

File: ohara-manager/client/src/components/Workspace/Settings/SettingsHooks.js (hunk 7 above)
Comment (eechih) [reply ohara-manager/client/src/components/Workspace/Settings/SettingsHooks.js:64]: Is this a workaround? If yes, that's ok.

**Comment by eechih** [comment]
> @eechih I've come up with a solution to allow dialog component to enable/disable the confirm button.

@konekoya 感謝～ 這個問題就先用這個解法就好，之後再安排時間把它改的更有彈性 :)

**Comment by konekoya** [comment]
> @konekoya 感謝～ 這個問題就先用這個解法就好，之後再安排時間把它改的更有彈性 :)

Sorry... I couldn't think of a better solution for now 😅

File: ohara-manager/client/src/components/Workspace/Settings/SettingsHooks.js (hunk 7 above)
Comment (konekoya) [reply ohara-manager/client/src/components/Workspace/Settings/SettingsHooks.js:64]: 🙏

**Comment by konekoya** [comment]
@eechih updates which address issues you mentioned in our earlier discussion :)

File: ohara-manager/client/src/components/Workspace/Settings/SettingsHooks.js [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+  const sections = [
+    {
+      heading: 'Topics',
+      components: [
+        {
+          icon: <SubjectIcon />,
+          title: 'Topics in this workspace',
+          type: SETTINGS_COMPONENT_TYPES.PAGE,
+          componentProps: {
+            children: (
```
Comment (konekoya) [inline ohara-manager/client/src/components/Workspace/Settings/SettingsHooks.js:130]: @eechih This is where you can supply your own React component

File: ohara-manager/client/src/components/Workspace/Settings/SettingsStyles.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+          ? theme.palette.background.paper
+          : 'none'};
+        width: 680px;
+        margin: auto;
+        padding: ${theme.spacing(12, 3)};
+        height: 100%;
+      }
+
+      @media (max-width: 1200px) {
+        width: 1200px;
```
Comment (konekoya) [inline ohara-manager/client/src/components/Workspace/Settings/SettingsStyles.js:44]: Fixed the layout issue. Now we display a scrollbar under 1200px width. This is not align with Mui's breakpoint though, we might consider tweaking this later on...

**Comment by konekoya** [comment]
retry manager-e2e

**Comment by konekoya** [comment]
retry manager-e2e

**Comment by konekoya** [comment]
Since the workspace settings pages are being heavily rewritten. The end-to-end tests involved are disabled for now. cc @oharastream/frontend

**Pull request merged**